VM/rel2vm_params.py

```diff
@@ -0,0 +1,1124 @@
+
+    if faultprop.Mw < 3.5:
+        optimise = False
+        logger.info(f"No optimisation for {srf_meta['name']} : mw<3.5")
+
+    if land0 >= target_land_coverage:
+        logger.info(
+            f"No optimisation for {srf_meta['name']} : land coverage >= {target_land_coverage}%"
+        )
+        optimise = False
+
+    # modify VM if necessary
+    if optimise:
+        logger.info(
+            f"Optimising : {srf_meta['name']} has mw>= 3.5 and land coverage < {target_land_coverage}%"
+        )
+
+        # rotation based on centre line bearing at this point
+        l1 = centre_lon(vm0_region[2])
+        l2 = centre_lon(vm0_region[3])
+        mid = geo.ll_mid(l1, vm0_region[2], l2, vm0_region[3])
+        bearing = round(geo.ll_bearing(mid[0], mid[1], l2, vm0_region[3]))
+
+        # wanted distance is at corners, not middle top to bottom
+        _, xlen1, ylen1 = determine_vm_extent(
+            rjb,
+            hh,
+            srf_meta["corners"].reshape((-1, 2)),
+            rot=bearing,
+            wd=ptemp,
+        )
+
+        logger.debug(
+            f"Pre-optimisation origin: {origin}, bearing: {bearing}, xlen: {xlen1}, ylen: {ylen1}"
+        )
+        # cut down ocean areas
+        origin, bearing, xlen1, ylen1, = reduce_domain(
+            origin,
+            bearing,
+            xlen1,
+            ylen1,
+            hh,
+            ptemp,
+            logger=logger,
+        )
+        logger.debug(
+            f"After optimisation origin: {origin}, bearing: {bearing}, xlen: {xlen1}, ylen: {ylen1}"
+        )
+
+        try:
+            c1, c2, c3, c4 = geo.build_corners(origin, bearing, xlen1, ylen1)
+        except ValueError:
+            error_message = f"Error for vm {srf_meta['name']}. Raising exception."
+            logger.log(qclogging.NOPRINTCRITICAL, error_message)
+            raise ValueError(error_message)
+        else:
+            logger.debug(f"Optimised corners of the VM are: {(c1, c2, c3, c4)}")
+
+        # proportion in ocean
+        land1 = get_vm_land_proportion(c1, c2, c3, c4, wd=ptemp)
+        logger.debug(f"Optimised land coverage found to be {land1}%")
+
+        # adjust region to fit new corners
+        plot_region = (
+            min(c4[0], c1[0], c3[0], c2[0], plot_region[0]),
+            max(c4[0], c1[0], c3[0], c2[0], plot_region[1]),
+            min(c4[1], c1[1], c3[1], c2[1], plot_region[2]),
+            max(c4[1], c1[1], c3[1], c2[1], plot_region[3]),
+        )
+    else:  # not optimised
+        # store original variables as final versions
+        ylen1 = ylen0
+        xlen1 = xlen0
+        land1 = land0
+        c1, c2, c3, c4 = o1, o2, o3, o4
+    # not enough land in final domain
+    if math.floor(land1) == 0:
+        logger.info(
+            "Land coverage is less than 1%. Setting xlen and ylen to 0, not creating VM"
+        )
+        xlen1 = 0
+        ylen1 = 0
```

Comment:
Setting these to 0 is a hangover for when processing multiple VMs so that there was appropriate output so that the generation didn't stop when it found an invalid VM.

Reply:
Yeah, but this causes the vm_params dict to not be generated later on. So we may as well 'return None' here instead of calculating zlen which will only get discarded

Reply:
I will return None if xlen1==0 or ylen1==0 or zlen==0.
Then optimised results can be unindented.

Reply:
I fixed a bit - welcome your feedback.

Reply:
Hang on- then do we just return None when it's too deep?
```
    if fault_depth > rrup:
        if deep_rupture:
            pass
        else:
            logger.warning(
                "fault_depth > rrup. Fault too deep, will not generate VM. "
                f"Setting xlen, ylen to 0. Previous values: x:{xlen0}, y:{ylen0}"
            )
            xlen0 = 0
            ylen0 = 0
```
Should we just return None here?
```
    if fault_depth > rrup and not deep_rupture:
        logger.warning(
            "fault_depth > rrup. Fault too deep, will not generate VM. "
            f"Setting xlen, ylen to 0. Previous values: x:{xlen0}, y:{ylen0}"
        )
        return None
```

Reply:
You could, however you'd need the srf.path file to be written before then so it's still available for plotting